### PR TITLE
Add description of render delay when user interacting and mitigation

### PR DESCRIPTION
**Pull request by michaellisitsa**
- [Add] Docs: Explain render delay after user interaction and mitigation strategies

### CODE REVIEW

File: docs/interactive-diagram-best-practices.md [hunk 1]
```diff
@@ -98,3 +98,46 @@ The toggle-panels example provides examples of how to deal with this inter-depen
...
+![Screenshot of user diagram](_media/interactive-diagram-best-practices/user-interaction-complex-states.png)
+
+</div>
+
+#### Handle multiple events
+
+During the time that ClearCalcs is processing params, the `getStoredParams` function will still return the old values of `storedParams` i.e. before the click/drag event. That means if multiple events happen in quick succession, then a queue of `setStoredParams` may form each being unaware of its other changes. There are some advanced techniques to mitigate these issues, to be used with caution:
+
+-   prevent user interaction until render has been called.
+-   use DOM state rather than `getStoredParams` to calculated new `storedParams`.
```
Comment (JulianGoh5) [inline docs/interactive-diagram-best-practices.md:143]: Minor typo on the "calculated", should be present tense

**Review by JulianGoh5 — APPROVED** [review]
Looks great! Found a few nitpicky things, will leave up to you as to whether you want to review now or later

File: docs/interactive-diagram-best-practices.md [hunk 2]
```diff
@@ -98,3 +98,46 @@ The toggle-panels example provides examples of how to deal with this inter-depen
...
+
+<div style="text-align: center;">
+
+![Screenshot of user diagram](_media/interactive-diagram-best-practices/user-interaction-complex-states.png)
+
+</div>
+
+#### Handle multiple events
+
+During the time that ClearCalcs is processing params, the `getStoredParams` function will still return the old values of `storedParams` i.e. before the click/drag event. That means if multiple events happen in quick succession, then a queue of `setStoredParams` may form each being unaware of its other changes. There are some advanced techniques to mitigate these issues, to be used with caution:
```
[the remark this answers is not in the document]
Comment (michaellisitsa) [reply docs/interactive-diagram-best-practices.md]: Yes agree, the whole section about multiple events needed more detail, so have added a flowchart and more detail t o clarify. See d8493c385278bc889ae2265b9eb5a851b43ed4c3

File: docs/interactive-diagram-best-practices.md (hunk 1 above)
Comment (michaellisitsa) [reply docs/interactive-diagram-best-practices.md]: Updated in d8493c385278bc889ae2265b9eb5a851b43ed4c3

File: docs/interactive-diagram-best-practices.md [hunk 3]
```diff
@@ -98,3 +98,46 @@ The toggle-panels example provides examples of how to deal with this inter-depen
...
+ClearCalcs manages the timing of receiving user interaction data and calling render. There is a minor delay where ClearCalcs receives the serialized message from the Interactive Diagram's iframes, stores its value and sends another message with all the `params` and the newly updated `storedParams` for the diagram to render.
+
+The delay may vary based on a number of factors:
+
+-   diagram sends/receives large chunks of data.
+-   client is busy performing other computations e.g. solving other maths on the calculator.
+-   client is on a low power device.
+-   ClearCalcs needs to perform server-side computations to re-calculate `params`.
+
+Below is a simple flow showing a number being clicked and changing its colour to blue. The number may show a delayed switch to blue, but this behaviour is acceptable. There is a minor risk, if the diagram is particularly complex, that the user may try to click again before render is called again. Some advanced strategies are presented in [Handling Duplicate Interactions](/interactive-diagram-best-practices?id=handle-multiple-events)
```
[the remark this answers is not in the document]
Comment (michaellisitsa) [reply docs/interactive-diagram-best-practices.md]: Updated in d8493c385278bc889ae2265b9eb5a851b43ed4c3

File: docs/interactive-diagram-best-practices.md [hunk 4]
```diff
@@ -98,3 +98,46 @@ The toggle-panels example provides examples of how to deal with this inter-depen
 
 ?> Important: the `setStoredParams()` function should only ever be called in response to user interaction, never on `params` changing or other non-user initiated event.
 
+## User Interaction UX
```
[the remark this answers is not in the document]
Comment (michaellisitsa) [reply docs/interactive-diagram-best-practices.md]: Simplifed heading in d8493c385278bc889ae2265b9eb5a851b43ed4c3

**Comment by michaellisitsa** [comment]
@JulianGoh5 Ready for re-review. Particularly new flowchart and updated section around multiple user interactions.

File: docs/interactive-diagram-best-practices.md [hunk 5]
```diff
@@ -98,3 +98,56 @@ The toggle-panels example provides examples of how to deal with this inter-depen
...
+
+<div style="text-align: center;">
+
+![Screenshot of user diagram](_media/interactive-diagram-best-practices/user-interaction-complex-states.png)
+
+</div>
+
+#### Handle multiple events
+
+In [examples/toggle-panels](https://github.com/ClearCalcs/custom-diagram-boilerplate/blob/main/examples/toggle-panels), a user can click on multiple panels and the toggle state of every panel is stored in an array. If a user interacts clicks on a 2nd panel while ClearCalcs is still processing the 1st array received from the diagram, the storedParams from the first click may be lost.
```
Comment (JulianGoh5) [inline docs/interactive-diagram-best-practices.md:140]: > If a user interacts clicks on a 2nd panel

File: docs/interactive-diagram-best-practices.md [hunk 6]
```diff
@@ -98,3 +98,56 @@ The toggle-panels example provides examples of how to deal with this inter-depen
...
+
+The reason for this is that `getStoredParams` function is not instantly updated as soon as `setStoredParams` is called, rather only after Cleasrcalcs has finished processing and calls render again. Any user interactions in the meantime will get the `storedParams` i.e. before the click/drag event. See flowchart:
```
Comment (JulianGoh5) [inline docs/interactive-diagram-best-practices.md:142]: > Cleasrcalcs

Comment (michaellisitsa) [reply docs/interactive-diagram-best-practices.md]: Dang. Fixed 4f2bfe5e1b980146b083cb33642b8456a5f8c57a

File: docs/interactive-diagram-best-practices.md (hunk 5 above)
Comment (michaellisitsa) [reply docs/interactive-diagram-best-practices.md]: Doh Fixed in 4f2bfe5e1b980146b083cb33642b8456a5f8c57a

**Pull request merged**